lib/src/solver/failure.dart

```diff
@@ -93,39 +92,22 @@ class _Writer {
   String write() {
     var buffer = StringBuffer();
 
-    // SDKs whose version constraints weren't matched.
-    var sdkConstraintCauses = <Sdk>{};
-
-    // SDKs implicated in any way in the solve failure.
-    var sdkCauses = <Sdk>{};
-
-    for (var incompatibility in _root.externalIncompatibilities) {
-      var cause = incompatibility.cause;
-      if (cause is PackageNotFoundCause) {
-        var sdk = cause.sdk;
-        if (sdk != null) {
-          sdkCauses.add(sdk);
-        }
-      } else if (cause is SdkCause) {
-        sdkCauses.add(cause.sdk);
-        sdkConstraintCauses.add(cause.sdk);
-      }
-    }
-
     // If the failure was caused in part by unsatisfied SDK constraints,
     // indicate the actual versions so we don't have to list them (possibly
     // multiple times) in the main body of the error message.
     //
     // Iterate through [sdks] to ensure that SDKs versions are printed in a
     // consistent order
-    var wroteLine = false;
-    for (var sdk in sdks.values) {
-      if (!sdkConstraintCauses.contains(sdk)) continue;
-      if (!sdk.isAvailable) continue;
-      wroteLine = true;
-      buffer.writeln('The current ${sdk.name} SDK version is ${sdk.version}.');
+    final notices = _root.externalIncompatibilities
+        .where((c) => c.cause.notice != null)
+        .map((c) => c.cause.notice)
+        .whereNotNull()
```

Comment:
Am I misunderstanding, or you do double null-element elimination here?

```suggestion
        .map((c) => c.cause.notice)
        .whereNotNull()
```

Reply:
good catch